DZ2: Buttons

cjb opened the pull request:
@keybase/react-hackers 

Here are the button changes for DZ2!  Stylesheet in the app is updated.  It's duplicating the Zeplin style guide buttons page:

<img width="944" alt="screen shot 2016-02-19 at 11 55 38 am" src="https://cloud.githubusercontent.com/assets/21217/13182814/9e368108-d701-11e5-9e37-73ac0d0e3cb0.png">

Please review, thanks!

chrisnojima commented:
- [ ] divs needs to be justify-contents = flex-start and align-items = flex-start
- [ ] secondary text needs to be black

cjb commented:
@chrisnojima Thanks, fixed!  And rebased onto master.

<img width="879" alt="screen shot 2016-02-19 at 1 27 09 pm" src="https://cloud.githubusercontent.com/assets/21217/13184905/85e5a5b0-d70c-11e5-8eac-b7c0afc15947.png">

gabriel commented on shared/common-adapters/button.js.flow:
```diff
@@ -7,11 +7,17 @@ export type Props = {
   style?: ?Object,
   primary?: ?bool,
+  secondary?: ?bool,
+  follow?: ?bool,
```
Should we use an enum (string) for these button types?

I don't think we should have types here for following/unfollow

Also why ?bool and not bool  ... we dont want these values to be true,false,null

chrisnojima replied:
Follow buttons are special

On Friday, February 19, 2016, Gabriel Handford <notifications@github.com>
wrote:

> In shared/common-adapters/button.js.flow
> <https://github.com/keybase/client/pull/2070#discussion_r53529669>:
>
> > @@ -7,11 +7,17 @@ export type Props = {
> >    label: ?string,
> >    style?: ?Object,
> >    primary?: ?bool,
> > +  secondary?: ?bool,
> > +  follow?: ?bool,
>
> Should we use an enum (string) for these button types?
>
> I don't think we should have types here for following/unfollow
>
> Also why ?bool and not bool ... we dont want these values to be
> true,false,null
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/keybase/client/pull/2070/files#r53529669>.
>

gabriel replied:
ahh ok

MarcoPolo replied:
I agree that these should be enums, because it doesn't make sense to be able to set danger: true, and secondary: true.

gabriel commented on shared/common-adapters/button.desktop.js:
```diff
@@ -31,11 +118,18 @@ Button.propTypes = {
   label: React.PropTypes.string,
   style: React.PropTypes.object,
   primary: React.PropTypes.bool,
+  secondary: React.PropTypes.bool,
```
can you import the type from flow and nuke the prop types definition so we don't have duplicate defs?

MarcoPolo commented:
:+1: % use enums and :fire: proptypes :skull:

gabriel commented:
:ship: :shipit:

cjb commented:
I switched to using a type enum in 88859e5, so now it's `<Button type='Primary' />` instead of `<Button primary />`.  It's kind of unsatisfying:

* I had to change the tracker1 render path too.
* Had to change every instance of `<Button />` to use the new type.  The type is now not-optional, so every `<Button />` with no type becomes `<Button type='Secondary' />`.
* The type is now case-sensitive where it wasn't before, though in theory flow should catch incorrect cases.  In practice it isn't catching them, @MarcoPolo is checking out why.

cjb commented:
Okay, as part of fixing up flow with @MarcoPolo's help, I added flow tags to files that weren't being tested (shared/update, shared/more/style-sheet) and fixed some types!

Ready for a final look?

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -2,40 +2,124 @@
-  small: React.PropTypes.bool,
-  more: React.PropTypes.bool
+  render () {
+    if (!flags.tracker2) {
```
in other components we've been adding a dz2 prop, we might want to do that here

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -2,40 +2,124 @@
+    let smallStyle = {}
+
+    // First apply styles for the main button types.
+    switch (this.props.type) {
```
maybe break this out into a helper, see text.desktop.js

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -47,13 +131,79 @@ const buttonCommon = {
+  color: globalColorsDZ2.white,
+  borderRadius: 55,
+  fontSize: 16,
+  height: '32px',
```
can just be height: 32

chrisnojima replied:
note: the things where you do have to put a unit: https://facebook.github.io/react/tips/style-props-value-px.html

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -47,13 +131,79 @@ const buttonCommon = {
+  buttonFollowing: {
+    ...buttonCommon2,
+    backgroundColor: globalColorsDZ2.white,
+    border: 'solid 2px ' + globalColorsDZ2.green,
```
prefer backtick templating
```
`solid 2px ${globalColorsDZ2.green}`
```

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -47,13 +131,79 @@ const buttonCommon = {
+    minWidth: '125px'
+  },
+  buttonSmall: {
+    height: '28px',
```
height: 28

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -47,13 +131,79 @@ const buttonCommon = {
+    backgroundColor: globalColorsDZ2.blue,
+    disabledOpacity: 0.2,
+    marginRight: 10,
+    minWidth: '125px'
```
minWidth: 125

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -47,13 +131,79 @@ const buttonCommon = {
+    backgroundColor: globalColorsDZ2.green,
+    disabledOpacity: 0.3,
+    marginRight: 10,
+    minWidth: '125px'
```
minWidth: 125

chrisnojima commented on shared/common-adapters/button.js.flow:
```diff
@@ -6,12 +6,13 @@ export type Props = {
+  disabled?: ?bool,
   small?: ?bool,
-  more?: ?bool
+  more?: ?bool,
```
maybe we can deprecate the more flag from the button class. its not in the style guide

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -2,40 +2,124 @@
+    return (
+      <FlatButton
+        onClick={this.props.onClick}
+        style={{...backgroundStyle, ...this.props.style, ...smallStyle}}
```
why smallStyle after this.props.style,  i assume you want this.props.style last so it actually overrides everything

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -2,40 +2,124 @@
+        style={{...backgroundStyle, ...this.props.style, ...smallStyle}}
+        labelStyle={{...styles2.buttonLabel, ...labelStyle}}
+        label={this.props.label || this.props.more && '•••'}
+        primary={this.props.type === 'Primary'}
```
this feels weird to me, we should enforce that its primary or secondary

cjb replied:
> we should enforce that its primary or secondary

Hm, why?  'alert' is an example of a type that isn't primary or secondary, and material-ui doesn't depend on having one of primary/secondary set.

cjb commented:
@chrisnojima Thanks!  Did everything except enforcing primary or secondary, that one doesn't make sense to me -- want to come over when you're free to talk about it?

chrisnojima commented on shared/common-adapters/button.desktop.js:
```diff
@@ -2,40 +2,127 @@
+      case 'Primary':
+        backgroundStyle = {
+          ...styles2.buttonPrimary,
+          opacity: this.props.disabled ? styles2.buttonPrimary.disabledOpacity : 1
```
lets move this out and not have it 10 times

chrisnojima commented:
nah we're good :+1:
